pkg/controller/instance/plan_execution_engine.go

```diff
@@ -1,332 +1,228 @@
+// (serial/parallel). Task execution might result in success, error and fatal error. It is to distinguish between transient
+// and fatal errors.  Transient errors are retryable, so the corresponding Plan/Phase are still in progress:
+//  └── first-operator-zljnmj
+//     └── Plan deploy (serial strategy) [IN_PROGRESS]
+//        └── Phase main [IN_PROGRESS]
+//           └── Step everything (ERROR)
+//
+// However, this does not apply to fatal errors! Should a  fatal error occur, we will, in the spirit of "fail-loud-and-proud",
+// abort current execution, resulting in a plan status like:
+//  └── first-operator-zljnmj
+//     └── Plan deploy (serial strategy) [FATAL_ERROR]
+//        └── Phase main [FATAL_ERROR]
+//           └── Step everything (FATAL_ERROR)
+//
+// Furthermore, a transient ERROR during a step execution, means that the next step may be executed if the step strategy
+// is "parallel". In case of a fatal error, it is returned alongside with the new plan status and published on the event bus.
+func executePlan(pl *activePlan, em *engtask.EngineMetadata, c client.Client, enh engtask.KubernetesObjectEnhancer) (*v1alpha1.PlanStatus, error) {
+	if pl.Status.IsTerminal() {
+		log.Printf("PlanExecution: Plan %s for instance %s is terminal, nothing to do", pl.name, em.InstanceName)
+		return pl.PlanStatus, nil
 	}
 
-	// we don't want to modify the original state -> need to do deepcopy
-	newState := plan.PlanStatus.DeepCopy()
+	planStatus := pl.PlanStatus.DeepCopy()
+	planStatus.Status = v1alpha1.ExecutionInProgress
 
-	// render kubernetes resources needed to execute this plan
-	planResources, err := prepareKubeResources(plan, newState, metadata, enhancer)
-	if err != nil {
-		return newState, err
-	}
+	phasesLeft := len(pl.spec.Phases)
+	// --- 1. Iterate over plan phases ---
+	for _, ph := range pl.spec.Phases {
+		phaseStatus := getOrCreatePhaseStatus(ph.Name, planStatus)
 
-	// do a next step in the current plan execution
-	allPhasesCompleted := true
-	for _, ph := range plan.spec.Phases {
-		currentPhaseState, _ := getPhaseFromStatus(ph.Name, newState)
-		if isFinished(currentPhaseState.Status) {
-			// nothing to do
-			log.Printf("PlanExecution: Phase %s on plan %s and instance %s is in state %s, nothing to do", ph.Name, plan.name, metadata.instanceName, currentPhaseState.Status)
+		// Check current phase status: skip if finished, proceed if in progress, break out if a fatal error has occurred
+		if isFinished(phaseStatus.Status) {
+			phasesLeft = phasesLeft - 1
 			continue
-		} else if isInProgress(currentPhaseState.Status) {
-			newState.Status = v1alpha1.ExecutionInProgress
-			currentPhaseState.Status = v1alpha1.ExecutionInProgress
-			log.Printf("PlanExecution: Executing phase %s on plan %s and instance %s - it's in progress", ph.Name, plan.name, metadata.instanceName)
+		} else if isInProgress(phaseStatus.Status) {
+			phaseStatus.Status = v1alpha1.ExecutionInProgress
+		} else {
+			break
```

Comment:
at this point no fatal error could occured, right? so we can never fall into the break part? Should this switch rather be AFTER we apply steps?